source/fsicalmanagement/business/password_hashing_service.d

```diff
@@ -0,0 +1,98 @@
+module fsicalmanagement.business.password_hashing_service;
+
+/**
+ * Provides functionality to hash passwords and validate passwords against
+ * hashes.
+ */
+interface PasswordHashingService
+{
+    /**
+     * Generates a hash from a $(D_PARAM password).
+     * Params:
+     * password = The $(D_PARAM password) to hash.
+     *
+     * Returns: The generated hash.
+     */
+    string generateHash(const string password) const @safe;
+
+    /**
+     * Validates $(D_PARAM password) against a given $(D_PARAM hash).
+     * Params:
+     * password = The password to validate.
+     * hash = The hash to validate against.
+     *
+     * Returns: Whether or not the validation was successful.
+     */
+    bool checkHash(const string password, const string hash) const @safe;
+}
+
+/**
+ * A stub implementation of `PasswordHashingService` which does not actually do
+ * any hashing.
+ */
+class StubPasswordHashingService : PasswordHashingService
+{
+    /**
+     * Simply returns the given $(D_PARAM password).
+     *
+     * Params:
+     * password = The $(D_PARAM password) to return.
+     *
+     * Returns: $(D_PARAM password).
+     */
+    string generateHash(const string password) const @safe pure nothrow
+    {
+        return password;
+    }
+
+    /**
+     * Validates that $(D_PARAM password) is equal to $(D_PARAM hash).
+     * Params:
+     * password = The password to validate.
+     * hash = The hash to validate against.
+     *
+     * Returns: Whether or not $(D_PARAM password) and $(D_PARAM hash) are
+     *          equal.
+     */
+    bool checkHash(const string password, const string hash) const @safe pure nothrow
+    {
+        return password == hash;
+    }
+}
+
+/**
+ * A $(LINK2 https://github.com/Abscissa/DAuth, dauth) based implementation of
+ * `PasswordHashingService` using SHA256 as hashing algorithm. Hashes are
+ * provided in a $(LINK2 https://en.wikipedia.org/wiki/Crypt_%28C%29, crypt(3))
+ * compatible form.
+ */
+class SHA256PasswordHashingService : PasswordHashingService
+{
+    import dauth : dupPassword, isSameHash, makeHash, parseHash;
```

Comment:
According to its docs, one should favor `toPassword(uchar[])` over `dupPassword`, as `string`s are not reliably nullable, which could impose a security risk.

Reply:
Unfortunately, this means that `uchar[]` bubbles up all the way. I am not sure if this is a good idea.

Reply:
It seems as if `dauth` also provides a `Password` type (`RefCounted!PasswordData`, thich itself is a struct wrapping `ubyte[]`) that is to be used for these occassions, then only the function taking the password from the UI would need to take `char[]`, and also the intent of the variable gets more clear from its type.

Reply:
But this means, that we tie ourselves very closely to dauth, which I‘d like to avoid. I‘ll have a look at how things are, if we use char[] all the time. Maybe it is not that bad.

Reply:
@pheerai addressed in 89b2b95. I used `char[]` as the type for passwords in all occasions.

As mentioned in my previous comment, I don't want to use DAuth's `Password` type in general, because it binds us to DAuth more closely.

Alsways using `char[]` is fine I guess and other implementations might also want to zero the passwords, once not needed anymore.

Any comments?

Reply:
If working with `char[]` is fine it's ok.

Even though I don't understand why we cannot use `Password` and, when in doubt or changing the implementation, just `alias Password = char[]`, or write a small wrapper struct ourselves. This would give us the additional comfort without too much effort in case of a change, plus it'll make the intent of the array more clear.